lightwood/encoders/time_series/helpers/transformer_helpers.py

```diff
@@ -0,0 +1,78 @@
+import math
+import torch
+import torch.nn as nn
+
+
+def len_to_mask(lengths, zeros):
+    # Clean trick from:
+    # https://stackoverflow.com/questions/53403306/how-to-batch-convert-sentence-lengths-to-masks-in-pytorch
+    mask = torch.arange(lengths.max())[None, :] < lengths[:, None]
+    if zeros:
+        mask = ~mask  # Logical not
+    return mask
+
+
+class PositionalEncoding(nn.Module):
+    def __init__(self, d_model, dropout=0.2, max_len=5000):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(
+            torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model)
+        )
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        self.register_buffer("pe", pe)
+
+    def forward(self, x):
+        x = x + self.pe[: x.size(0), :]
+        return self.dropout(x)
+
+
+class TransformerModel(nn.Module):
+    def __init__(self, ninp, nhead, nhid, nlayers, dropout=0.2):
+        super(TransformerModel, self).__init__()
+        self.src_mask = None
+        # Lezcano: This could be an embedding if the data is in a range [0, R-1]
+        self.encoder = nn.Linear(1, ninp)
+        self.pos_encoder = PositionalEncoding(ninp, dropout)
+        encoder_layers = nn.TransformerEncoderLayer(ninp, nhead, nhid, dropout)
+        self.transformer_encoder = nn.TransformerEncoder(encoder_layers, nlayers)
+        self.decoder = nn.Linear(ninp, 1)
+        self.init_weights()
+
+    def _generate_square_subsequent_mask(self, sz):
+        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
+        # Take the logarithm
+        mask = (
+            mask.float()
+            .masked_fill(mask == 0, float("-inf"))
+            .masked_fill(mask == 1, float(0.0))
+        )
```

Comment:
I'm not quite sure I understand the purpose of this code.

To me it seems like it generates a matrix of `sz`x`sz` where the "lower" triangle + the diagonal are `0` and the "upper" triangle is `-inf` but I fail to see how this mask is helpful/relevant.

Maybe add a comment explaining how it works here or above the forward call to the `transformer_encoder` where it is used.

But @paxcema is the expert here, so maybe this is super obvious and I'm missing something

Reply:
This is part of the transformer architecture for processing time series. The attention in the transformer, a priori, does not take in account position (think of attention in images). This triangular mask is for the attention, so that words in a sentence can just attend to previous words. This, in turn, enables decoding, as the model that you train has just learnt to attend to previous positions.